src/ILCompiler.ReadyToRun/src/JitInterface/CorInfoImpl.ReadyToRun.cs

```diff
@@ -888,170 +925,537 @@ private void getCallInfo(ref CORINFO_RESOLVED_TOKEN pResolvedToken, CORINFO_RESO
+            // function pointer
+            else if (targetMethod.HasInstantiation)
+            {
+                pResult->kind = CORINFO_CALL_KIND.CORINFO_VIRTUALCALL_LDVIRTFTN;  // stub dispatch can't handle generic method calls yet
+                pResult->nullInstanceCheck = true;
+            }
+            // Non-interface dispatches go through the vtable.
+            else if (!targetMethod.OwningType.IsInterface)
+            {
+                pResult->kind = CORINFO_CALL_KIND.CORINFO_VIRTUALCALL_VTABLE;
+                pResult->nullInstanceCheck = true;
 
-                    if (pResult->exactContextNeedsRuntimeLookup)
+                // We'll special virtual calls to target methods in the corelib assembly when compiling in R2R mode, and generate fragile-NI-like callsites for improved performance. We
+                // can do that because today we'll always service the corelib assembly and the runtime in one bundle. Any caller in the corelib version bubble can benefit from this
+                // performance optimization.
+                MethodDesc callerMethod = HandleToObject(callerHandle);
+                if (!MethodInSystemVersionBubble(callerMethod) || !MethodInSystemVersionBubble(targetMethod))
+                {
+                    pResult->kind = CORINFO_CALL_KIND.CORINFO_VIRTUALCALL_STUB;
+                }
+            }
+            else
+            {
+                // Insert explicit null checks for cross-version bubble non-interface calls. 
+                // It is required to handle null checks properly for non-virtual <-> virtual change between versions
+                pResult->nullInstanceCheck = callVirtCrossingVersionBubble && !targetMethod.OwningType.IsInterface;
+                pResult->kind = CORINFO_CALL_KIND.CORINFO_VIRTUALCALL_LDVIRTFTN;
+
+                // We can't make stub calls when we need exact information
+                // for interface calls from shared code.
+
+                if (// If the token is not shared then we don't need a runtime lookup
+                    pResult->exactContextNeedsRuntimeLookup
+                    // Handle invalid IL - see comment in code:CEEInfo::ComputeRuntimeLookupForSharedGenericToken
+                    && entityFromContext(pResolvedToken.tokenContext) is MethodDesc methodDesc && methodDesc.IsSharedByGenericInstantiations)
+                {
+                    ComputeRuntimeLookupForSharedGenericToken(DictionaryEntryKind.DispatchStubAddrSlot, ref pResolvedToken, null, targetMethod, ref pResult->codePointerOrStubLookup);
+                }
+                else
+                {
+                    // We use an indirect call
+                    pResult->codePointerOrStubLookup.constLookup.accessType = InfoAccessType.IAT_PVALUE;
+                    pResult->codePointerOrStubLookup.constLookup.addr = null;
+                }
+            }
+
+            pResult->hMethod = ObjectToHandle(methodToCall);
+
```

Comment:
@trylek In the re-write we lost this line. This is probably the reason that the Release build is failing.
```
pResult->accessAllowed = CorInfoIsAccessAllowedResult.CORINFO_ACCESS_ALLOWED;
```